internal/configwatcher/watcher.go

```diff
@@ -0,0 +1,109 @@
+type ConfigWatcher struct {
+	Config  config.Config
+	PubSub  pubsub.PubSub
+	Tracer  trace.Tracer `inject:"tracer"`
+	subscr  pubsub.Subscription
+	msgTime time.Time
+	done    chan struct{}
+	startstop.Starter
+	startstop.Stopper
+}
+
+// ReloadCallback is used to tell others that the config has changed.
+// This gets called whenever it has actually changed, but it might have
+// changed because we were told about it in pubsub, so we don't publish
+// a message if the hashes are the same.
+func (cw *ConfigWatcher) ReloadCallback(cfgHash, rulesHash string) {
+	ctx := context.Background()
+	ctx, span := otelutil.StartSpanMulti(ctx, cw.Tracer, "ConfigWatcher.ReloadCallback", map[string]any{
+		"new_config_hash": cfgHash,
+		"new_rules_hash":  rulesHash,
+	})
+	defer span.End()
+
+	// don't publish if we have recently received a message (this avoids storms)
+	now := time.Now()
+	if now.Sub(cw.msgTime) < time.Duration(cw.Config.GetGeneralConfig().ConfigReloadInterval) {
+		otelutil.AddSpanField(span, "sending", false)
+		return
+	}
+
+	message := now.Format(time.RFC3339)
+	otelutil.AddSpanFields(span, map[string]any{"sending": true, "message": message})
+	cw.PubSub.Publish(ctx, ConfigPubsubTopic, message)
+}
+
+// SubscriptionListener listens for messages on the config pubsub topic and reloads the config
+// if a new set of hashes is received.
+func (cw *ConfigWatcher) SubscriptionListener(ctx context.Context, msg string) {
+	_, span := otelutil.StartSpanWith(ctx, cw.Tracer, "ConfigWatcher.SubscriptionListener", "message", msg)
+	defer span.End()
+
+	// parse message as a time in RFC3339 format
+	msgTime, err := time.Parse(time.RFC3339, msg)
+	if err == nil {
+		return
+	}
+	cw.msgTime = msgTime
+	// maybe reload the config (it will only reload if the hashes are different,
+	// and if they were, it will call the ReloadCallback)
+	cw.Config.Reload()
+}
+
+// Monitor periodically wakes up and tells the config to reload itself.
+// If it changed, it will publish a message to the pubsub through the ReloadCallback.
+func (cw *ConfigWatcher) monitor() {
+	cw.done = make(chan struct{})
+	cfgReload := cw.Config.GetGeneralConfig().ConfigReloadInterval
+	// adjust the requested time by +/- 10% to avoid everyone reloading at the same time
+	reload := time.Duration(float64(cfgReload) * (0.9 + 0.2*rand.Float64()))
+	ticker := time.NewTicker(time.Duration(reload))
+	for {
+		select {
+		case <-cw.done:
+			return
+		case <-ticker.C:
+			cw.Config.Reload()
+		}
+	}
+}
+
+func (cw *ConfigWatcher) Start() error {
+	if cw.Tracer == nil {
+		cw.Tracer = noop.NewTracerProvider().Tracer("test")
+	}
+	if cw.Config.GetGeneralConfig().ConfigReloadInterval != 0 {
+		go cw.monitor()
+	}
+	cw.subscr = cw.PubSub.Subscribe(context.Background(), ConfigPubsubTopic, cw.SubscriptionListener)
+	cw.Config.RegisterReloadCallback(cw.ReloadCallback)
+	return nil
+}
+
+func (cw *ConfigWatcher) Stop() error {
+	cw.subscr.Close()
```

Comment:
Will this race with an ongoing `Reload` ? Maybe we can close `cw.done` first to make sure no code is going to utilize the pub/sub channel anymore